linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,455 @@
+                srw_lock_acquire_shared(threadpool->srw_lock);
+                TASK_RESULT curr_task_result = interlocked_compare_exchange(&threadpool->task_array[current_index].task_state, TASK_WORKING, TASK_WAITING);
+                srw_lock_release_shared(threadpool->srw_lock);
+                if (TASK_WAITING == curr_task_result)
+                {
+                    int64_t consume_pos = interlocked_compare_exchange_64(&threadpool->consume_idx, current_index+1, current_index);
+                    if (consume_pos == current_index)
+                    {
+                        srw_lock_acquire_shared(threadpool->srw_lock);
+                        THREADPOOL_TASK* temp_item = &threadpool->task_array[consume_pos];
+                        srw_lock_release_shared(threadpool->srw_lock);
+
+                        // Get the item and make sure it's valid
+                        if (temp_item != NULL)
+                        {
+                            temp_item->task_func(temp_item->task_param);
+                        }
+                        // I'm done with this item mark it as complete
+                        srw_lock_acquire_shared(threadpool->srw_lock);
+                        interlocked_exchange(&threadpool->task_array[consume_pos].task_state, TASK_NOT_USED);
+                        srw_lock_release_shared(threadpool->srw_lock);
+                    }
+                }
+            }
+        } while (interlocked_add(&threadpool->quitting, 0) != 1);
+        (void)interlocked_decrement(&threadpool->thread_count);
+    }
+}
+
+static int reallocate_threadpool_array(THREADPOOL* threadpool)
+{
+    int result;
+    {
+        // Lock the array here
+        srw_lock_acquire_exclusive(threadpool->srw_lock);
+
+        int32_t existing_count = interlocked_add(&threadpool->task_array_count, 0);
+
+        // Double the items
+        (void)interlocked_exchange(&threadpool->task_array_count, existing_count*2);
+
+        // Reallocate here
+        THREADPOOL_TASK* temp_array = realloc_2((void*)threadpool->task_array, existing_count*2, sizeof(THREADPOOL_TASK));
+        if (temp_array == NULL)
+        {
+            LogError("Failure realloc_2(threadpool->task_array: %p, threadpool->task_array_count: %" PRId32", sizeof(THREADPOOL_TASK): %zu", threadpool->task_array, existing_count*2, sizeof(THREADPOOL_TASK));
+            result = MU_FAILURE;
+        }
+        else
+        {
+            // Clear the newly allocated memory
+            memset(temp_array+existing_count, 0, existing_count*sizeof(THREADPOOL_TASK));
+            threadpool->task_array = temp_array;
+
+            // Normalize the consumed index
+            //(void)interlocked_exchange_64(&threadpool->consume_idx, threadpool->consume_idx % existing_count);
+            // Make the insert index at the new items that were allocated
+            (void)interlocked_exchange_64(&threadpool->insert_idx, existing_count+1);
```

Comment:
>e_64(&threadpool->insert_idx, existing_count+1); [](http://example.com/codeflow?start=37&length=48)

I can foresee some problems here.
Imagine that we have existing_count be 4.
We are happily consuming from index 2.
Entries 3, 0, 1, 2 all have tasks in them.
Again, consume index is 2.
Now we double, we have 8 items.
Items 0 to 3 inclusive are all used: { x, x, x, x, 0, 0, 0, 0 }

Produce index gets set to 4.
A new task is added at 4: { x, x, x, x, x, 0, 0, 0 }
Consuming continues happily at 2. So we eat up all tasks: { x, x, 0, 0, 0, 0, 0, 0 }
Consume index is 5 now.
Produce index is 5 also.

And the tasks 0 and 1 remain unconsumed until we have more things scheduled. Potentially for a very long time (if no scheduling happens for 1h, then things do not get executed for 1h.

That will not be good.